vk/middleware.go

```diff
@@ -49,50 +78,32 @@ func enableCors(ctx *Ctx, domain string) {
 	}
 }
 
-// generate a HandlerFunc that passes the request through a set of Middleware first and Afterware after
-func augmentHttpHandler(inner HandlerFunc, middleware []Middleware, afterware []Afterware) HandlerFunc {
-	return func(r *http.Request, ctx *Ctx) (interface{}, error) {
-		defer func() {
-			// run the afterware (which cannot affect the response)
-			// even if something in the request chain fails
-			for _, a := range afterware {
-				a(r, ctx)
+// ErrorMiddleware returns a middleware that wraps a handler.
+func ErrorMiddleware() Middleware {
+	return func(inner HandlerFunc) HandlerFunc {
+		return func(w http.ResponseWriter, r *http.Request, ctx *Ctx) error {
+			if err := inner(w, r, ctx); err != nil {
+				ctx.Log.ErrorString(fmt.Sprintf("ERROR: traceid: %s, msg: %s", ctx.RequestID(), err.Error()))
+
+				if e, ok := err.(Error); ok {
+					// we received a trusted error, which means we can pass on the status and message set on it.
+					w.WriteHeader(e.Status())
+					errJson, err := json.Marshal(e)
+					if err != nil {
+						return errors.Wrap(err, "could not marshal error into json")
```

Comment:
Do we need to expose this level of detail to the public or would it make sense to log this for debugging; opting instead to expose something a little more opaque like "an unexpected error occurred" to clients?